view/frontend/web/js/track-checkout.js

```diff
@@ -0,0 +1,70 @@
+/* Track checkout funnel transition in magento's default checkout */
+
+// TODO: support logged in user
+
+var currentStep = '';
+
+function waitFor(condition, doWhenReady) {
+    var checkCondition = function () {
+        if (condition()) {
+            doWhenReady();
+        } else {
+            setTimeout(checkCondition, 50);
+        }
+    };
+    checkCondition();
+}
+
+function trackFunnel(step) {
+    if (step === currentStep) {
+        // prevent duplicate tracking
+        return;
+    }
+    BoltTrack.recordEvent(step);
+    currentStep = step;
+}
+
+function setupListnerForShippingForm() {
+    var requiredFields = jQuery("#shipping-new-address-form .field._required");
+    var inputElements = [];
+    requiredFields.each(function(_, e) {
+        $elm = jQuery(e).find("input,select");
+        inputElements.push($elm);
+        $elm.change(function() {
+           var complete = true;
+           for (var i = 0; i < inputElements.length; i++) {
+               if (!inputElements[i].val() && inputElements[i].attr("name") !== "region") {
+                   complete = false;
+                   break;
+               }
+           }
+           if (complete) {
+               trackFunnel("onShippingAddressComplete");
+           }
+        });
+    });
+}
+
+function init() {
+    trackFunnel("onCheckoutStart");
+
+    waitFor(
+        // wait for shipping form to be fully rendered, ie., we have 9 required fields in the page.
+        function() { return jQuery("#shipping-new-address-form .field._required input,select").length === 9; },
```

Comment:
I would estimate that 1/3 of the merchants I've worked with customize their checkout to include new required fields or to remove some of the required fields. If they do, this won’t work on their sites.

Maybe you should use a required value range as the condition instead of a specific number. Something like this:

> function() { return jQuery("#shipping-new-address-form .field._required input,select").length >== 5; },

Reply:
I updated to `>= 9`. We'll working closely with first few beta merchant and hopefully we can find better approach during beta